### PR TITLE
added ZSH support

### PR DESCRIPTION
**Pull request by alg**
Hi there. I found that the original script doesn't work for ZSH autocompletion, so I adapted it.

### CODE REVIEW

**Comment by dalexj** [comment]
Hey, this would be great to include here for support.

In my bash version, I have cached the return of the small elixir script to grab all mix tasks, so it's not recompiling and executing every time you hit [tab], which can be annoying to wait for. Could you include this functionality in your script?

**Comment by alg** [comment]
Hey mate. Makes sense. I'll have a look in a minute.

**Comment by dalexj** [comment]
looking into the `compctl` docs, I did something very close to this

``` zsh
compctl -k "($(
    mix run -e '
      Mix.Task.load_all
        |> Enum.map(&(Mix.Task.task_name &1))
        |> Enum.sort
        |> Enum.join(" ")
        |> IO.puts
    '))" mix
```

which would rerun every time I `cd` as to not keep out of date with mix tasks switching between projects/etc

**Comment by alg** [comment]
If I understand it right, the cache should be expired when mix.exs timestamp changes. It'll handle both cases -- when you add more modules to your existing project and switching between projects. It won't handle the case when you add mix tasks by writing them, but we can think of something.

I'd like to try timestamp-based caching and see how it goes. What do you think?

**Comment by alg** [comment]
Also, I just hit the case when I tried auto-completing mix when there are errors in the project. It fails to compile the source and displays an ugly message instead of the completions list. Should probably handle this too next.

**Comment by dalexj** [comment]
that would be some nice improvement on top of what I already have. go ahead and take a stab at the timestamp-based approach

I'll get an issue for the case with errors in the project, can address that later

**Comment by alg** [comment]
Ok, ready for the review. I changed the plan a bit and replaced the mtime with md5 hash of the mix.exs. Should be pretty reliable.

Also, I noticed that if the project requires recompiling, everything that we see during the compilation ends up among the suggestions (file names, paths, parts of messages etc). So I added an extra compilation phase before running the mix tasks snippet and continuing judging by the exit code.

**Comment by dalexj** [comment]
hey, looking through the code here and it all seems correct. I like this version for caching much better.

however for some reason when I'm running this through my shell, I dont seem to be picking up the cache when I hit tab.

File: mix_autocomplete.zsh [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+
+
+function _mix() {
+  if [[ ! -e "mix.exs" ]]; then
+    return
+  fi
+
+  local mix_md5
+  local completions
+  local current_mix_md5=`md5 -q mix.exs`
```
Comment (msoedov) [inline mix_autocomplete.zsh]: I'm not sure that `md5` available on all os/platforms besides of OSX

Comment (alg) [inline mix_autocomplete.zsh]: Yup. I noticed that when testing on Ubuntu. It's called md5sum there. Will add a piece of code to find the appropriate tool. In the meantime, I tested it on fresh Ubuntu server install and it hits the cache just fine. Can it be some misconfig on your side?

**Comment by dalexj** [comment]
gotcha. I dont use zsh so I have no zshrc. I got the cache working by running these before the script

``` zsh
autoload -U compinit
compinit
zstyle ':completion::complete:*' use-cache on
```

**Comment by alg** [comment]
Updated with md5/md5sum selector. Still not sure what is wrong with cache on your side.

**Comment by alg** [comment]
> gotcha. I dont use zsh so I have no zshrc. I got the cache working by running these before the script

Sweet!

**Comment by dalexj** [comment]
ok looks good, I'll have to pull your added functionality in to my bash script now. thanks!

**Comment by alg** [comment]
cheers ;)

**Pull request merged**